### PR TITLE
Add OPO length-weighted reward baseline to GRPO

### LINKED ISSUE #3540: Implement On-Policy RL with Optimal re- ward baseline (OPO) (Microsoft Research)
### Method description

What is OPO ?

This paper introduces OPO (On-Policy RL with Optimal reward baseline), a simpler and more efficient algorithm that:

   * Uses exact on-policy training for greater stability and better exploration.

   * Derives an optimal reward baseline to minimize gradient variance.

   * Eliminates the need for value/reference models and regularization.

Key Benefits of OPO:

   * Theoretical Soundness: Minimizes gradient variance with an optimal baseline.

   * Training Stability: Achieves stable learning without KL/entropy penalties.

   * Better Performance: Outperforms baselines in math reasoning tasks with more diverse and less repetitive outputs.

Performs almost equivalent to GRPO without need for KL penalties.

PS : I would like to implement this as this would be a learning experience for me 🤗 

### Open source status

- [x] The method implementation is available
- [x] The model weights are available
- [x] The training datasets are available

### Provide useful links for the implementation

Paper : https://arxiv.org/pdf/2505.23585

Code : https://github.com/microsoft/LMOps/tree/main/opo

Base  Model : https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-7B

Training Data : https://huggingface.co/datasets/Skywork/Skywork-OR1-RL-Data

Evaluation Datasets:  MATH-500, AIME 2024 , and AIME 2025.

### PR DESCRIPTION
**Pull request by YAO-001**
# What does this PR do?

This PR adds the length-weighted reward baseline from On-Policy RL with Optimal Reward Baseline (OPO) as an opt-in option in `GRPOTrainer`.

As discussed in #3540, most of OPO is already expressible through the existing GRPO configuration:

- `beta=0.0` disables the reference model and KL regularization.
- `scale_rewards="none"` disables reward standard-deviation scaling.
- `num_iterations=1` uses fresh rollouts for every optimizer update.

The remaining algorithmic difference is OPO's group baseline:

$$
b^* = \frac{\sum_i L_i R_i}{\sum_i L_i}
$$

where $L_i$ and $R_i$ are the completion length and reward of response $i$.

### Changes

- Add `reward_baseline` to `GRPOConfig`:
  - `"mean"` preserves the existing GRPO behavior and remains the default.
  - `"length_weighted"` enables the OPO baseline.
- Compute the length-weighted baseline from completion lengths gathered across processes.
- Exclude unscorable completions from both the numerator and denominator.
- Reject `reward_baseline="length_weighted"` with `multi_objective_aggregation="normalize_then_sum"`, where the OPO scalar-reward baseline is not defined.
- Add an OPO entry and configuration example to `docs/source/paper_index.md`.
- Add deterministic coverage for unequal lengths, equal-length equivalence, partially unscorable groups, fully unscorable groups, configuration validation, and a complete training step.

The default GRPO behavior is unchanged, and this PR adds no new dependencies.

### Usage

```python
from trl import GRPOConfig

training_args = GRPOConfig(
    reward_baseline="length_weighted",
    scale_rewards="none",
    beta=0.0,
    num_iterations=1,
    entropy_coef=0.0,
)
```

### Validation

```text
python -m pytest tests/test_grpo_trainer.py \
  -k "reward_baseline or train_scale_rewards or train_normalize_then_sum_aggregation or warning_raised_all_rewards_none" \
  -q

9 passed, 176 deselected
```

The length-weighted training test runs through generation, reward calculation, advantage computation, and parameter updates with fixed completions and analytically computed expected advantages.

Fixes #3540

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. See #3540.
- [x] Did you make sure to update the documentation with your changes? Added the OPO entry and usage example to `paper_index.md`.
- [x] Did you write any new necessary tests? Added deterministic baseline and end-to-end training coverage.

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

cc @qgallouedec

---

> [!NOTE]
> **Medium Risk**
> Changes advantage computation in a core GRPO training path, but only when explicitly opting into `length_weighted`; default training is unchanged.
> 
> **Overview**
> Adds opt-in **`reward_baseline`** on `GRPOConfig` so `GRPOTrainer` can use OPO’s completion-length-weighted group baseline instead of the default arithmetic mean when `multi_objective_aggregation="sum_then_normalize"`.
> 
> **`length_weighted`** computes advantages from ∑(LᵢRᵢ)/∑Lᵢ per prompt group, using gathered completion lengths (or tool masks when present) and excluding unscorable completions from both numerator and denominator; **`normalize_then_sum`** is rejected for this mode. Default **`mean`** behavior is unchanged.
> 
> Documentation adds an OPO section in `paper_index.md` with a sample `GRPOConfig`, and tests cover config validation plus deterministic advantage values through a full training step.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e9cf80b9e702649d74a69db955c5ba3a209a7d19. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by YAO-001** [comment]
Hi @qgallouedec, thanks again for your guidance on #3540. The [Tests](https://github.com/huggingface/trl/actions/runs/30736786968) and [Build PR Documentation](https://github.com/huggingface/trl/actions/runs/30736787192) workflows for #6620 are currently waiting for maintainer approval.

When you have a moment, could you please approve these workflow runs so the CI checks can start? I have run the focused GRPO tests and the full pre-commit checks locally, and they pass. Thank you!

**Comment by qgallouedec** [comment]
Thanks a lot for this @YAO-001, and sorry for the slow reply. The implementation is clean and matches exactly what I described in #3540.

This one is on me though. I scoped it in the issue and invited a PR, and looking again at the demand I don't think GRPO should carry a permanent config field for it. reward_baseline adds a public option plus a cross-field constraint with multi_objective_aggregation, and a second branch in the advantage block that has to stay in sync with RLOO. The issue has had no follow-up interest since it was opened, so I'd rather not grow the config surface here.

Closing this one, thanks for the effort though
